RetroAchievements/RAWeb
# feat: async double write and metrics aggregation

Adding asynchronous double-write and metrics aggregation for games, player-games, player-sessions, player-achievements while preserving legacy features.

This change adds complexity and some boilerplate overhead but is very much worth maintaining so we can scale the platform better. 

`player_games` and `player_achievement_sets` tables have been updated:
```sh
php artisan migrate
```

> **Note**
> Setting `QUEUE_DRIVER` to `sync`, which is the default locally for new setups, will run the jobs synchronously as the driver name implies.
> Handling jobs synchronously has a major impact on execution time. 
> Handling jobs asynchronously, as it will happen on production, is done by running `./artisan horizon` when `QUEUE_DRIVER` is set to `redis`.
> Tests use the `sync` queue driver.

This implementation makes us of Laravel's Events, Listeners, Jobs, Commands, which eventually call re-usable Actions. 
Events dispatch queued listeners which in turn will dispatch Jobs with specific rate limiting behaviour.
Only in rare cases are Jobs dispatched directly.

To see how events cascade into listeners take a look at `app/Platform/EventServiceProvider`.

`dorequest` is heavily affected by this, thus tests have been updated to take new behaviour into account.

How to test:
- Validate expected data manipulation in database tables: `player_sessions`, `player_games`, `Awarded`, 'player_achievements', 'UserAccounts', 'GameData', `user_activities`, `Activity`
- Requests towards Connect API (`dorequest.php`), ideally end-to-end via emulators
- Reset progress via settings (`controlpanel.php`)
- Run provided commands, run any of the below without arguments and the `-h` option to see usage infos

```sh
./artisan ra:platform:player:unlock-achievement luchaos 1
./artisan ra:platform:player:unlock-achievement luchaos 1,2,3,4,5,6,8,9,15,16,17,18,19,20,21,22,23,24,25,32,38,60,129,130 --hardcore
./artisan ra:platform:player:update-game-metrics luchaos 1,228
./artisan ra:platform:game:update-metrics 1,228
./artisan ra:platform:player:update-game-metrics luchaos --outdated
```

> **Note**
> There are zombie entries in the `player_achievements` table due to progress resets that have not been taken into account yet and only removed unlocks from `Awarded` until now.
> To clean those efficiently I'm looking for help with that so i can focus on tuning the queues and bug-fixing in the meantime. 

Jobs implement various framework-native rate limiting contracts to control scaling.
Adjusting rate-limiting and locking jobs to be unique will be worth looking into. After we gained more insights on how the system behaves running the queues in parallel.
As it's running in parallel, double-writing, this will give us enough time to validate all data and behaviour before doing the switch.

> **Note**
> Jobs only take primitives as input arguments to make sure we do not run into deserialization issues with changes in models in the future.
